rpc/src/transfer.rs

```diff
@@ -0,0 +1,202 @@
+        let source_account_id_bytes = evm_address_to_account_id_bytes(source_address);
+        let source_account_id =
+            AccountId::decode(&mut &source_account_id_bytes[..]).unwrap_or_default();
+
+        //transfer amount: 52-67 bytes
+        let mut value_bytes = [0u8; 16];
+        value_bytes.copy_from_slice(&message_bytes[52..68]);
+        let value_128: u128 = u128::from_be_bytes(value_bytes);
+
+        //nonce: 68-71
+        let mut nonce_bytes = [0u8; 4];
+        nonce_bytes.copy_from_slice(&message_bytes[68..72]);
+        let nonce: Index = u32::from_be_bytes(nonce_bytes).into();
+
+        let address = match eth_recover(&signature, &message_bytes, &[][..]) {
+            Some(addr) => addr,
+            None => {
+                return Err(Error {
+                    code: ErrorCode::ServerError(RUNTIME_ERROR),
+                    message: "Failed to recover message signer.".into(),
+                    data: None,
+                })
+            }
+        };
+
+        //make sure that the signature is signed by source_address
+        if address != source_address {
+            return Err(Error {
+                code: ErrorCode::ServerError(RUNTIME_ERROR),
+                message: "Source address mismatch.".into(),
+                data: None,
+            });
+        }
+        //make sure source address has sufficient balance???
+        let account_basic = api.account_basic(&at, source_address).unwrap();
+        if account_basic.balance < value_128.into() {
+            return Err(Error {
+                code: ErrorCode::ServerError(RUNTIME_ERROR),
+                message: "Insufficient balance.".into(),
+                data: None,
+            });
+        }
+        //make sure use correct nonce
+        let real_nonce = match api.account_nonce(&at, source_account_id) {
+            Ok(nonce) => nonce,
+            Err(e) => {
+                return Err(Error {
+                    code: ErrorCode::ServerError(RUNTIME_ERROR),
+                    message: "Failed to get nonce.".into(),
+                    data: Some(format!("{:?}", e).into()),
+                })
+            }
+        };
+        if nonce != real_nonce {
+            return Err(Error {
+                code: ErrorCode::ServerError(RUNTIME_ERROR),
+                message: "Account nonce mismatch.".into(),
+                data: None,
+            });
+        }
+
+        //submit a unsigned extrinsics into transaction pool
+        let _ = api
+            .submit_unsigned_transaction(&at, message_bytes, signature_bytes)
+            .map_err(|e| Error {
+                code: ErrorCode::ServerError(RUNTIME_ERROR),
+                message: "Failed to submit unsigned extrinsics.".into(),
+                data: Some(format!("{:?}", e).into()),
+            });
+
+        //TODO how to handle result???
+
+        Ok(0)
+    }
+
+    fn transfer_nonce(&self, evm_addr: String) -> Result<u32> {
```

Comment:
Do we need to handle with or without 0x?

Reply:
also maybe a size check for the String?

Reply:
No need to check `0x` prefix, I will add bytes length check.